client/logs.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2021 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"strconv"
+	"time"
+)
+
+const (
+	logReaderSize = 4 * 1024
+)
+
+type LogsOptions struct {
+	// Function called to write a single log to the output (required).
+	WriteLog WriteLogFunc
+
+	// The list of service names to fetch logs for (nil or empty slice means
+	// all services).
+	Services []string
+
+	// Total number of logs to fetch (before following if calling FollowLogs).
+	// If nil, use Pebble's default (10). If negative, fetch all buffered logs.
+	// If set to zero when calling FollowLogs, write no logs before following.
+	NumLogs *int
+}
+
+type WriteLogFunc func(timestamp time.Time, service, message string) error
```

Comment:
That means adding a field to the log format is impossible. We should define an external type here and pass that into the function instead, so that clients can ignore those fields and remain forwards compatible.

Reply:
Good call -- fixed. And at that point the function signature is so simple `func (entry LogEntry) error` that I removed the type altogether.